Add Python 3 rewrite of arc_summary.py (#6873)

scotws opened the pull request:
### Description

Add new Python 3 script `arc_summary3.py` as a complete rewrite of `arc_summary.py` to display basic information on the ARC status and various other parameters. This is provided in addition - not as a replacement - to the existing `arc_summary.py` tool. See #6873 for a discussion of the reasoning behind adding a new version of this tool while keeping a legacy version as well.

New options:

        -g/--graph    - Display crude graphic representation of ARC status and quit
        -r/--raw      - Print all available information as minimally formatted list and quit
        -s/--section  - Print a single section. This supersedes -p/--page, which is kept for
                        backwards use but marked as DEPRECIATED

Adds new sections with information on the ZIL and SPL. 

We now notify the user if sections L2ARC and VDEV are skipped instead of failing silently; note VDEV caching is currently disabled and slated for possible removal (see source code). Adds information on the ZFS and SPL versions to the header.

The **-s/--section** option is intended to replace the page number system, which required the user to remember which page number was of interest. The -p/--page options are still supported, but marked as DEPRECIATED. Current legal sections are `arc archits dmu l2arc spl tunables vdev zil`. It should be easier now to add and modify sections.

The **-r/--raw** option is intended to work with other tools such as `grep`. It respects the -a/-d options (alternate output format / descriptions included) where possible. 

The output of the **-g/--graph** option is intended to give a quick, rough overview as a visual orientation. An example (Ubuntu 16.04 LTS x86_64 with 24 GB RAM, 8 GB ARC max, ZFS stock version 0.6.5.9-2 with `/home` as ZFS mirror pool immediately after starting _Civilization VI_ on Steam on otherwise quiet machine): 
```
        ARC: 3.0 GiB (37.5 %)  MFU: 610.5 MiB  MRU: 2.3 GiB
    +----------------------------------------------------------+
    |FFFFRRRRRRRRRRRRRRRRR                                     |
    +----------------------------------------------------------+
```
`F` is for MFU, `R` for MRU, and `O` is used for "other" if necessary (not present in this example). 

`arc_summary3.py` was developed for Python 3.5. This follows the version of Python currently installed in Ubuntu 16.04 LTS. Few systems will have Python 3.6 installed yet.

### Known issues

The new script is based on the same internal logic as the original, so any error or issue present there will probably show up here as well. For instance, the number of anonymous hits can be negative the way it is calculated in both scripts; they both simply hide any negative value.

This script will probably make a bunch of test suites unhappy where Python 3 is not included. There is no experience with this script under extreme conditions (for example ARC throttling).

### How Has This Been Tested?

There is a unittest script `test_arc_summary3.py` at https://gist.github.com/scotws/aaf5d9c9317081e249b664a371ec4907
Most testing was done in-tree, comparing the output to that of the current `arc_summary.py` version.

The L2ARC section has **not seen any real-world use** because I do not have access to a L2ARC device on my machine.

### Other 

Switching to Python 3 results in a noticeably smaller file size despite the addition of several new features. Output of `wc` for both scripts:
```
    837    2586   28021 arc_summary3.py
   1020    2593   35538 arc_summary.py
```

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux code style requirements.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

scotws commented on cmd/arc_summary/test_arc_summary3.py:
```diff
@@ -0,0 +1,116 @@
+#!/usr/bin/env python3
```
Added a monkey-see, monkey-do test that is basically a straight copy of the one for arc_summary.py with the new parameters, and included it in Makefile and the `linux.run`. Don't see a way to automatically trigger the Python 3 test script test_arc_summary3.py included in `cmd/arc_summary`? That would be nice to do it additionally, because it can test individual functions. Would it make sense to move that script somewhere in the test folder as well?

behlendorf replied:
> Would it make sense to move that script somewhere in the test folder as well?

Yes, your going to need to and then add a simply wrapper which calls it with `log_must`.

scotws commented:
Uh, I just noticed that I didn't include a line for arc_summary3.py in `cmd/arc_summary/Makefile.am`. Should I add that in the same way as the present line for arc_summary.py?

behlendorf commented:
> Should I add that in the same way as the present line for arc_summary.py?

Yes, you can simply be able to add it to the `dist_bin_SCRIPTS` line and it'll be installed under `/bin/` and included in the standard packages.

scotws commented:
So I moved the Python test file over to the other ksh test routines and added a ksh call. Python is really, really unhappy about putting the test that far away from the actual program, so I had to do some trickery with the imports in the test routine, which then annoys flake8 and pylint (so I expect those tests to fail). Unless I'm missing something, we can have one or the other -- the test routine with the other tests or happy linters.

behlendorf commented on tests/zfs-tests/tests/functional/cli_user/misc/test_arc_summary3.py:
```diff
@@ -0,0 +1,122 @@
+
+import os
+import sys
+FILE_LOCATION = "../../../../../../cmd"
```
Because the ZTS can be run from either the installed `zfs-tests` package or in-tree it's not safe to assume the location of arc_summary3.  All the ZTS knows for certain is that by adding `arc_summary3.py` to `tests/zfs-tests/include/commands.cfg` (which you should do) the utility will be in the default `PATH`.  Perhaps the calling script can pass in the full path?

scotws commented:
Including arc_summary3.py in `.../commands.cfg`

AFAIK, the linters will fail as long as the imports don't come first, so any manipulation of the paths is going to make them unhappy. One unorthodox, brute-force way of dealing with the whole problem would be to create a symbolic link to the original arc_summary3.py in the same folder as the test (`.../cli_user/misc/`). -- Is there any way to just turn off the linting of certain files, a `.gitignore` for the testing suite?

behlendorf commented:
We're using `flake8` for a linter and it appears it supports disabling warning with inline annotations.  That should allow you to make it happy.

http://flake8.pycqa.org/en/latest/user/violations.html#ignoring-violations-with-flake8

scotws commented:
@behlendorf Thanks for pointing out inline flake8 stuff, testing that now. Will figure out PATH next.

scotws commented:
Importing PATH into test_arc_summary3.py now works on my machine in-tree when simulated with a manipulated PATH. Added comments to explain the trickery involved and why we turned off flake8 for the test program. Fingers crossed.

behlendorf commented on tests/runfiles/linux.run:
```diff
@@ -408,7 +408,7 @@ tests = ['zdb_001_neg', 'zfs_001_neg', 'zfs_allow_001_neg',
     'zpool_replace_001_neg', 'zpool_scrub_001_neg', 'zpool_set_001_neg',
     'zpool_status_001_neg', 'zpool_upgrade_001_neg', 'arcstat_001_pos',
-    'arc_summary_001_pos', 'dbufstat_001_pos']
+    'arc_summary_001_pos', 'arc_summary3_001.pos', 'dbufstat_001_pos']
```
Did you intentionally omit `arc_summary3_002.pos`?

behlendorf commented on cmd/arc_summary/Makefile.am:
```diff
@@ -1 +1,2 @@
 dist_bin_SCRIPTS = arc_summary.py
+dist_bin_SCRIPTS = arc_summary3.py
```
Since `__init__.py` needs to be included in `make dist` you'll need to add and `EXTRA_DIST` line to this Makefile.am.

```diff
+EXTRA_DIST = __init__.py
```

Can we include a comment somewhere, either in  `__init__.py` or the `Makefile.am` describing why this empty file is needed.

scotws replied:
Sorry, hadn't thought of adding file to `test/runfiles/linux.run`. Still a bit hazy on the structure of the test suit ... 

As for __init__.py, let's let this run through the test suite again without it -- I think we might not need it after all, but because I'm having trouble duplicating the suite here, I wasn't sure. Thanks again for your help!

rlaager commented:
A few of the doc strings for functions in arc_summary3.py are full sentences but don't end in periods.

Other than the above, the code _looks_ good on a read-through.

Running the plain command (arc_summary vs. arc_summary3), I get similar results. The numbers are rounded to one place rather than two, but that seems fine to me and I assume is intentional. I ran it on systems with and without L2ARC.

The L2ARC writes aren't being abbreviated. I'm not sure if that's intentional.
**arc_summary:**
```
L2 ARC Writes:
	Writes Sent:			100.00%	207.69k
```
**arc_summary3:**
```
L2ARC writes:
        Writes sent:                                    100 %     207670
```

scotws commented:
@rlaager, thanks for checking, especially the L2ARC stuff; I'll try to get to the fixes this week. -- @behlendorf at this point, I would suggest we just drop the Python test program, because it's adding too much complexity for too little value; even if we get it to work, nobody is going to be able to figure it out later. I'd include a link to the test program so it's not lost, and then we just go with the "native" ZFS test suite? -- Thanks again!

behlendorf commented:
@scotws sounds good.  I agree, the test program isn't worth the bother in this case.  Thanks.

gmelikov commented on tests/zfs-tests/tests/functional/cli_user/misc/test_arc_summary3.py:
```diff
@@ -0,0 +1,126 @@
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+# SUCH DAMAGE.
```
@scotws What is the license of this file? Maybe use CDDL too, as you did in arc_summary3_002_pos.ksh?

scotws replied:
@gmelikov Oops, thought I had, will take a look (but it'll be the weekend before I can get to any of this). Note that we've decided to remove that whole file from the tree because it's causing more trouble than it is worth; I'll keep it my github page with a link in the patch notes.

scotws commented:
Sorry, I'm having problems understanding what is going wrong with the tests, for instance 

`Makefile:1449: recipe for target 'shellcheck' failed`

with buildbot/Ubuntu 17.10 x86_64 (STYLE). Did I delete some file by mistake or leave a line in I shouldn't have?

behlendorf commented:
Sorry about the shellcheck warnings, they should be addressed today by #7040 and are unrelated to this change.  The version of shellcheck was updated on the builder and it included additional checks.

scotws commented:
If I understand the logs right, a bunch of stuff is failing because it can't find `/usr/lib/python2.7/subprocess.py` on the test machine - which is weird, because I have Ubuntu 16.04 here and I definitely have that file ...

behlendorf commented on cmd/arc_summary/Makefile.am:
```diff
@@ -1 +1,2 @@
 dist_bin_SCRIPTS = arc_summary.py
+dist_bin_SCRIPTS = arc_summary3.py
```
You need to append `arc_summary3.py` to `dist_bin_SCRIPTS` otherwise `arc_summary.py` won't be included in the package which is why the test suite can find it and those tests are failing.

```
dist_bin_SCRIPTS = arc_summary.py arc_summary3.py
```

behlendorf commented on tests/runfiles/linux.run:
```diff
@@ -408,7 +408,8 @@ tests = ['zdb_001_neg', 'zfs_001_neg', 'zfs_allow_001_neg',
     'zpool_status_001_neg', 'zpool_upgrade_001_neg', 'arcstat_001_pos',
-    'arc_summary_001_pos', 'arc_summary_002_neg', 'dbufstat_001_pos']
+    'arc_summary_001_pos', 'arc_summary_002_neg', 
+    'arc_summary3_001.pos', 'arc_summary3_002.pos', 'dbufstat_001_pos']
```
The test case is named `arc_summary3_001_pos` not `arc_summary3_001.pos`.  The `.ksh` will be added to the test framework if needed.  Plus these is no `arc_summary3_002.pos` test case in this PR.  This is why you're getting the verification failures.

scotws commented:
@behlendorf Argh, apols, I had completely forgotten about linux.run ... as far as I understand the failure logs now, Python 3 is missing on the target machines (here for Ubuntu 16.04):

```
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_user/misc/arc_summary3_001_pos (run as buildbot) [00:00] [FAIL]
22:27:10.26 ASSERTION: arc_summary3.py generates output and doesn't return an error code
22:27:10.26 /usr/bin/env: 'python3': No such file or directory
22:27:10.26 ERROR: eval arc_summary3.py  > /dev/null exited 127
```
Not sure what we want to do here -- have the test routine test for Python 3 and skip itself it isn't installed, or install Python 3 on the test machines? We could have the program see if it is running Python 2 and quit with a fake success code, but seems sort of dishonest and not very helpful to the user. (Sorry, I won't be able to get anything done here before the weekend.)

behlendorf commented:
@scotws I think we want to do three things.

1) Add `python3` to `tests/zfs-tests/include/commands.cfg`, this way if it's installed on the system it will be available in the test environments sanitized path.

```diff
diff --git a/tests/zfs-tests/include/commands.cfg b/tests/zfs-tests/include/comm
index e2d0c41..f2abb38 100644
--- a/tests/zfs-tests/include/commands.cfg
+++ b/tests/zfs-tests/include/commands.cfg
@@ -88,6 +88,7 @@ export SYSTEM_FILES='arp
     ps
     pwd
     python
+    python3
     quotaon
     readlink
     rm
```

2) As you suggested we should skip the test when `python3` isn't available.  This way it'll show up in the test report as SKIP instead of PASS or FAIL. There are few other tests which do something similar when the functionality they need isn't available for some reason.  The way to do this is to add a check like this to the beginning of the test case.

```sh
if python3 -V >/dev/null 2>&1; then
        log_unsupported "python3 is not installed"
fi
```

3) When the test images next get updated we can install python3 if the distribution provides packages.

scotws commented:
@behlendorf If I'm reading this correctly, **Amazon 2 x86_64** is failing because ... it's not even getting the rpm packages assembled? and **CentOS 6 x86_64** does have Python 3, but some version before Python 3.5, _which isn't even the newest version_. I'll see if if my ksh skills are good enough to figure out how to do so test to the effect of of `Python3 -V` followed by `(( $? < 3.5 ))` ...

scotws commented:
@behlendorf @rlaager Phew! As far as I can tell, the Amazon problem is something not related to my code? -- And thanks for all the help!

behlendorf commented:
@scotws it looks like the Amazon 2 issue is because by python3 isn't included in their main repository or installed by default.  It is available in an `amazon-linux-extras` repository, I'll get this installed in to the test image which should resolve the issue.

 http://build.zfsonlinux.org/builders/Amazon%202%20x86_64%20Release%20%28TEST%29/builds/343/steps/shell_2/logs/install

dinatale2 commented:
@scotws The PR to install python 3 on the builders was just merged. Would you mind refreshing this PR to launch another set of builds?

scotws commented:
@dinatale2 it seems to be missing something: `E: Unable to locate package libasan1
program finished with exit code 100` for some of these.

behlendorf commented:
@scotws the issues with the bots should be resolved, I've resubmitted the failed builds and I'll see about giving this another review.  Thanks!

scotws commented:
@behlendorf I'm not sure what is failing here, the Linux build? I didn't change anything :smiley: !

behlendorf commented:
@sckobras actually everything is passing which is great, the "Kernel.org Built-in"  builder isn't your fault and needs to be addressed in a different PR.  I'd like to do some manual testing with this before merging, but it looks very close to ready.  @gmelikov @rlaager time permitting it would be great if you could kick the tires a bit too.

rlaager commented on cmd/arc_summary/arc_summary3.py:
```diff
@@ -0,0 +1,833 @@
+    by the different "short scale" and "long scale" representations in
+    English, which use the same words for different values. See
+    https://en.wikipedia.org/wiki/Names_of_large_numbers and
+    https://physics.nist.gov/cuu/Units/prefixes.html .
```
I'd probably leave off the period on optionally add a colon after "and".

rlaager commented on cmd/arc_summary/arc_summary3.py:
```diff
@@ -0,0 +1,833 @@
+
+
+def section_archits(kstats_dict):
+    """Print information on how the caches are accessed ("arc hits")
```
I'd consider this a full sentence and use a period.

rlaager commented on cmd/arc_summary/arc_summary3.py:
```diff
@@ -0,0 +1,833 @@
+
+
+def section_dmu(kstats_dict):
+    """ Collection information on the DMU"""
```
"Collect" instead of "Collection"

rlaager commented on cmd/arc_summary/arc_summary3.py:
```diff
@@ -0,0 +1,833 @@
+
+
+def section_l2arc(kstats_dict):
+    """ Collect information on L2ARC device if present. If not, tell user
```
Stray leading space

rlaager commented on cmd/arc_summary/arc_summary3.py:
```diff
@@ -0,0 +1,833 @@
+
+
+def section_zil(kstats_dict):
+    """ Collect information on the ZFS Intent Log. Some of the information
```
Stray leading space

rlaager commented:
I threw in some comments about minor typographical things. While those ideally should be cleaned up, they're minor, and I changed my review to Approve so I'm not blocking this.

behlendorf commented on cmd/arc_summary/arc_summary3.py:
```diff
@@ -0,0 +1,833 @@
+
+    command = ["/sbin/modinfo", request, "-0"]
+    info = subprocess.run(command, stdout=subprocess.PIPE,
+                          universal_newlines=True)
```
On Centos 7.4, running python 3.4.5, I'm hitting the following error when running `arc_summary3.py` without arguments.  My guess is this  is caused by running zfs in-tree without installing kmods on the system. 

Can you update this function to read the module version from `/sys/module/{spl,zfs}/version` instead.  This will always work and is authoritative since it's the version of the module currently loaded in to the kernel.  That may differ from what `modinfo` reports if new kmods have been built but not yet loaded.

```
Traceback (most recent call last):
  File "./cmd/arc_summary/arc_summary3.py", line 833, in <module>
    main()
  File "./cmd/arc_summary/arc_summary3.py", line 790, in main
    print_header()
  File "./cmd/arc_summary/arc_summary3.py", line 354, in print_header
    zfs = get_version('zfs')
  File "./cmd/arc_summary/arc_summary3.py", line 327, in get_version
    modinfo = subprocess.run(command, stdout=subprocess.PIPE,
AttributeError: 'module' object has no attribute 'run'
```

scotws replied:
@behlendorf If this is my friend CentOS with Python 3.4, then the problem is that there it is using an old version of subprocess (seriously, what is it with CentOS?). See https://docs.python.org/3/library/subprocess.html , where it says "New in version 3.5". We could add a check at the beginning to make sure that it's 3.5 or later and fail gracefully; the good news is that this problem should go away by itself in a few months.

I assume we still want to check `/sys/module` and not `modinfo`? Will try to get to that over the weekend, but could be early next week. Thanks for the help!

behlendorf replied:
> I assume we still want to check /sys/module and not modinfo?

Yes, and it looks like `get_description` and `get_version` are the only callers ` subprocess` so perhaps we can avoid this compatibility issue entirely.

> the good news is that this problem should go away by itself in a few months.

Why is that?

scotws replied:
I'll see what I can do about `subprocess` next week ... I remember there was a work-around. -- I'm assuming that CentOS at some point will be upgrading Python 3 -- we're almost at Python 3.7, so 3.4 is pretty ancient.

behlendorf replied:
@scotws does look like CentOS provides both python 3.4 and 3.6 from the EPEL repository.  But if at all possible let's try and do something compatible with both.

scotws replied:
@behlendorf Sorry to bother you with this, but I can't seem to find a way to get the descriptions for the individual parameters from the `/sys/module/{spl,zfs}/` tree - only through `modinfo`. What am I missing? (Possibly this was the reason that the original `arc_summary.py` used `modinfo` for its version of `get_description`). Thanks!

behlendorf replied:
You're right, it looks like only the version is provided.  Well frankly these strings never change so while not ideal it wouldn't be terrible to hardcode a meaningful string in the utility.  The critical information is the version which is available.

scotws replied:
@behlendorf If we're going to hardcode the descriptions, we might as well make them better than they are now, with the real types (AFAIK, the bools are currently being displayed as ints). If possible, we could also include something along the lines of "less is better" or "should be 0". I'd just need somebody to proofread those very carefully who actually understands what is going on under the hood here ...

behlendorf replied:
Sorry, I badly misunderstood your comment before and I hope I didn't send you on a wild goose chase.  Initially I thought this simply refered to the module description itself, I see that's not the case you're talking about the parameter descriptions too.  You're right, that case isn't so simple and we don't want to hardcode that information since it'll be a significant maintence burden.  I'd be OK with any of the following solutions, with a preference for the first one.

* Update `get_description()` to work with python 3.4.

* Add `Requires: python >= 3.5` to spec file and somehow express in the script's shebang that at least version 3.5 needs to be used.  I'm not sure if that's possible since `/usr/bin/python3` doesn't seem to always link to the _newest_ python3 version, and it's possible to have `python34` and `python36` concurrently installed.

* Drop this functionality for now until we find the best way to handle it.

rlaager replied:
As I understand it, the issue is that `subprocess.run()` doesn't exist on older versions of Python 3. The fix is trivial. Use the perfectly good older equivalents instead. Do this instead:
```
info = subprocess.check_output(command)
raw_output = info.decode(sys.stdin.encoding).split('\0')
```

rlaager replied:
Full disclosure: `subprocess.run()` says it uses the `io.TextIOWrapper` default encoding, which says it uses `locale.getpreferredencoding(False)`. If you wanted to match more closely, you could use that instead of `sys.stdin.encoding` in my example. On my Ubuntu system, though, `sys.stdin.encoding` is `UTF-8` (from my terminal) and `locale.getpreferredencoding(False)` is `ANSI_X3.4-1968`. I have always used `sys.stdin.encoding` (or `stdout`) and it's always been fine for me. Using `sys.stdin.encoding` seems like the superior option.

scotws replied:
@rlaager Thanks for the suggestions! There are actually two issues at the moment: 

The first one is the `subprocess.run()` problem, which affects a very small number of machines -- if I'm seeing this correctly, it is only some CentOS versions. This will go away by itself as the last distributions catch up to Python 3.5, so I was actually planning to keep the current, high-level code and throw an `IF/THEN` version check to the effect of `< 3.5` around it with the older `subprocess.check_output()` for the stragglers as you suggest. When everybody is at 3.5 in a few months, we simply remove the `IF/THEN`. Yes, this is crude, but keeps the recommended, cleaner code for most machines, and is temporary anyway.

The second issue is that @behlendorf has pointed out we should be reading stuff from `/sys/module/{spl,zfs}/version` instead of using `modinfo` (like the original `arc_summary.py`) so we report what's actually in the kernel instead of what has been built but maybe not loaded (which I had honestly not thought of). That's also trivial, except that there seems no way to get the parameters descriptions from `/sys/module`. 

My current plan there is, if the user wants descriptions,  we load the parameters from `/sys/module` and use them as keys to a dictionary of parameters/descriptions we get from `modinfo`. Yes, that runs the risk that some descriptions might change between loaded/built-but-not-loaded, but that should be small. -- Unless I've missed something and there is some other way to retrieve the descriptions from `/sys/modules`, which would of course be best.

Thanks again!

behlendorf replied:
>  When everybody is at 3.5 in a few months.

I wouldn't get your hopes up for CentOS 7.  Since it's an enterprise distribution they're particuarly adverse to change and it wouldn't shock me if python 2.7 remains the default python version until it's [end-of-life in 2024](https://wiki.centos.org/About/Product).

The good news is that python 3.6 is also provided by the distribution, worst case we can do something distribution specific here and require `#!/bin/python36`.  But if we can fallback gracefully to something compatible that would be better.

richardelling replied:
personally, I don't see check_output going away anytime soon, but...

there no need to check versions, check to see if you can import run from subprocess thusly:
```python
info = ''
if 'run' in dir(subprocess):
  info = subprocess.run(...)
elif 'check_output' in dir(subprocess):
  info = subprocess.check_output(...)
```

scotws replied:
@richardelling Thanks, hadn't thought of that, that's pretty much what I did now.

scotws commented:
@behlendorf Sorry, took longer than I thought to get around to this. Could you please run it on your CentOS system again, I tried to test it for `subprocess.check_output()` as far as possible, but that's not like the real thing. Version now taken from `/sys/module`, descriptions still with `modinfo` because we can't get them from there. Thanks!

behlendorf commented on cmd/arc_summary/arc_summary3.py:
```diff
@@ -0,0 +1,849 @@
+    descs = {}
+    target_prefix = 'parm:'
+
+    # We would prefer to do this with /sys/modules -- see the discusion at
```
s/discusion/discussion

behlendorf commented on cmd/arc_summary/arc_summary3.py:
```diff
@@ -0,0 +1,849 @@
+        raw_output = info.stdout.split('\0')
+    else:
+        info = subprocess.check_output(command, universal_newlines=True)
+        raw_output = info.split('\0')
```
Almost there!  Everything now works as intended when the kmods exist. However we still throw an ugly exception when they're missing.  Can you just handle this a bit more gracefully and not print an error.  This case can legitimately come up for at least two reasons.

* ZFS was built direclty in to the kernel and not as an independant kmod, or
* a developer is using a custom kmod which was built in the zfs srouce in-tree and not loaded with modprobe


```
modinfo: ERROR: Module spl not found.
Traceback (most recent call last):
  File "./cmd/arc_summary/arc_summary3.py", line 849, in <module>
    main()
  File "./cmd/arc_summary/arc_summary3.py", line 843, in main
    section_calls[section](kstats)
  File "./cmd/arc_summary/arc_summary3.py", line 693, in section_spl
    descriptions = get_descriptions('spl')
  File "./cmd/arc_summary/arc_summary3.py", line 305, in get_descriptions
    info = subprocess.check_output(command, universal_newlines=True)
  File "/usr/lib64/python3.4/subprocess.py", line 617, in check_output
    raise CalledProcessError(retcode, process.args, output=output)
subprocess.CalledProcessError: Command '['/sbin/modinfo', 'spl', '-0']' returned non-zero exit status 1
```

behlendorf commented on cmd/arc_summary/arc_summary3.py:
```diff
@@ -0,0 +1,849 @@
+            print(format_raw_line(name, value))
+
+    # Tunables and SPL must be handled separately because they come from a
+    # different source and have desciptions the user might request
```
s/desciptions/descriptions

scotws commented:
@behlendorf Good point. I've thrown a `try/except` check around the `subprocess` call, note that neither `pylint3` nor `flake8` seem to know `CalledProcessError` though it's perfectly ligit (see [https://docs.python.org/2/library/subprocess.html#subprocess.CalledProcessError](url) and [https://docs.python.org/3.5/library/subprocess.html#subprocess.CalledProcessError](url) ) so they're probably going to be unhappy. I'd suggest we actually terminate the program if we can't access kmod because that might a sign of a different, deeper error the user should probably know about.

loli10K commented:
@scotws i think we should be using `subprocess.CalledProcessError`, not just `CalledProcessError`:

```diff
diff --git a/arc_summary3.py b/arc_summary3.py
index 21b57e9..adcb781 100644
--- a/arc_summary3.py
+++ b/arc_summary3.py
@@ -308,7 +308,7 @@ def get_descriptions(request):
             info = subprocess.check_output(command, universal_newlines=True)
             raw_output = info.split('\0')
 
-    except CalledProcessError:
+    except subprocess.CalledProcessError:
         print("Error: Descriptions not available (can't access kernel module)")
         sys.exit(1)
```

Follow-up question: is it ok to `print()` errors on the standard output? Would it be better to `sys.stderr.write(error + "\n")` instead?

scotws commented:
@loli10K Well, when you put it like that, it seems painfully obvious 😃. Thanks, flake8 is happy now, too. -- As for `print()`, my usage scenario for `arc_summary3.py` is a quick, interactive tool and not something you'd, say, include in a shell script with redirects. For that, you'd want a more machine-friendly output, and then I'd separate the streams (insert Ghostbuster reference here). Maybe machine-friendly output of some data would be a future option?

scotws commented:
@behlendorf Yay! And thanks again for all the help.
